File: question/neetcode/binary_tree/binary-tree-right-side-view.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* node, int d, vector<int>& a){
        if(not node) return;
        
        if(a.size()==d) a.push_back(node->val);
        else a[d] = node->val;

        dfs(node->left,d+1,a);
        dfs(node->right,d+1,a);
    }

    vector<int> rightSideView(TreeNode* root) {
        vector<int> a;

        dfs(root,0,a);

        return a;
    }
};

File: question/neetcode150/trees/binary-tree-right-side-view.py
from typing import Optional, List
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        q = deque()
        ans = []

        if root:
            q.append(root)

        while q:
            ans.append(-1)
            for _ in range(len(q)):
                n = q.popleft()
                ans[-1] = n.val
                if n.left:
                    q.append(n.left)
                if n.right:
                    q.append(n.right)

        return ans


